src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -60,80 +103,458 @@ W3DDisplay::W3DDisplay() :
+                W3D::Shutdown();
+                GameMath::Shutdown();
+                throw CODE_07;
+                captainslog_debug("Unable to set render device");
+            }
+        }
+
+        if (g_theGameLODManager->Get_Static_LOD_Level() == STATLOD_INVALID) {
+            g_theGameLODManager->Set_Static_LOD_Level(g_theGameLODManager->Find_Static_LOD_Level());
+        } else {
+            if (g_theWriteableGlobalData->m_textureReductionFactor > 0) {
+                W3D::Set_Texture_Reduction(g_theWriteableGlobalData->m_textureReductionFactor, 32);
+                g_theGameLODManager->Set_Texture_Reduction_Factor(g_theWriteableGlobalData->m_textureReductionFactor);
+            }
+        }
+
+        if (g_theWriteableGlobalData->m_gammaValue != 1.0f) {
+            Set_Gamma(g_theWriteableGlobalData->m_gammaValue, 0.0f, 1.0f, false);
+        }
+
+        Init_Assets();
+        Init_2D_Scene();
+        Init_3D_Scene();
+        W3DShaderManager::Init();
+
+        m_nativeDebugDisplay = new W3DDebugDisplay();
+        m_debugDisplay = m_nativeDebugDisplay;
+
+        if (m_nativeDebugDisplay != nullptr) {
+            m_nativeDebugDisplay->Init();
+            GameFont *font;
+
+            if (g_theGlobalLanguage && g_theGlobalLanguage->Debug_Display_Font().Name().Is_Not_Empty()) {
+                font = g_theFontLibrary->Get_Font(g_theGlobalLanguage->Debug_Display_Font().Name(),
+                    g_theGlobalLanguage->Debug_Display_Font().Point_Size(),
+                    g_theGlobalLanguage->Debug_Display_Font().Bold());
+            } else {
+                font = g_theFontLibrary->Get_Font("FixedSys", 8, false);
+            }
+
+            m_nativeDebugDisplay->Set_Font(font);
+            m_nativeDebugDisplay->Set_Font_Width(13);
+            m_nativeDebugDisplay->Set_Font_Height(9);
+        }
+
+        // DX8WebBrowser::Initialize
+        m_initialized = true;
+
+        if (g_theWriteableGlobalData->m_displayDebug) {
+            m_debugDisplayCallback = Stat_Debug_Display;
+        }
+    }
 }
 
 // 0x0073D030
 void W3DDisplay::Reset()
 {
     Display::Reset();
-
     auto *iter = s_3DScene->Create_Iterator(false);
+
     for (iter->First(); !iter->Is_Done(); iter->Next()) {
         auto *renderObj = iter->Current_Item();
         renderObj->Add_Ref();
         s_3DScene->Remove_Render_Object(renderObj);
         renderObj->Release_Ref();
     }
-    s_3DScene->Destroy_Iterator(iter);
 
+    s_3DScene->Destroy_Iterator(iter);
     m_isClippedEnabled = false;
-
     s_assetManager->Release_Unused_Assets();
 
     if (g_theWriteableGlobalData != nullptr) {
         g_theWriteableGlobalData->m_drawSkyBox = false;
     }
 }
 
+void Draw_Graphical_Framerate_Bar()
+{
+    static int lastTime = rts::Get_Time();
+    int time = rts::Get_Time();
+    float f1 = 1000.0f / (float)(time - lastTime) / (1000.0f / (float)g_theWriteableGlobalData->m_framesPerSecondLimit);
+
+    if (f1 > 1.0f) {
```

Comment:
this is a clamp. `f1 = std::clamp(f1, 0.0f, 1.0f);`

Reply:
fixed.